onnxruntime/core/providers/rocm/nn/pool.cc

```diff
@@ -0,0 +1,353 @@
+POOLING_KERNEL_VERSIONED_WITH_INDICES(MaxPool, MLFloat16, MaxPool<8>, 11, 11)
+POOLING_KERNEL_WITH_INDICES(MaxPool, float, MaxPool<8>, 12)
+POOLING_KERNEL_WITH_INDICES(MaxPool, double, MaxPool<8>, 12)
+POOLING_KERNEL_WITH_INDICES(MaxPool, MLFloat16, MaxPool<8>, 12)
+POOLING_KERNEL_WITH_INDICES(MaxPool, int8_t, MaxPool<8>, 12)
+POOLING_KERNEL_WITH_INDICES(MaxPool, uint8_t, MaxPool<8>, 12)
+
+GLOBAL_POOLING_KERNEL(GlobalAveragePool, float, AveragePool, 1)
+GLOBAL_POOLING_KERNEL(GlobalAveragePool, double, AveragePool, 1)
+GLOBAL_POOLING_KERNEL(GlobalAveragePool, MLFloat16, AveragePool, 1)
+GLOBAL_POOLING_KERNEL(GlobalMaxPool, float, MaxPool<1>, 1)
+GLOBAL_POOLING_KERNEL(GlobalMaxPool, double, MaxPool<1>, 1)
+GLOBAL_POOLING_KERNEL(GlobalMaxPool, MLFloat16, MaxPool<1>, 1)
+
+class MiopenPoolingDescriptor final {
+ public:
+  MiopenPoolingDescriptor() : desc_(nullptr) {
+  }
+
+  ~MiopenPoolingDescriptor() {
+    if (desc_ != nullptr) {
+      miopenDestroyPoolingDescriptor(desc_);
+      desc_ = nullptr;
+    }
+  }
+
+  MiopenPoolingDescriptor(const MiopenPoolingDescriptor&) = delete;
+  MiopenPoolingDescriptor& operator=(const MiopenPoolingDescriptor&) = delete;
+
+  Status Set(miopenPoolingMode_t mode,
+             const gsl::span<const int64_t>& kernel_shape,
+             const gsl::span<const int64_t>& pads,
+             const gsl::span<const int64_t>& strides) {
+    if (!desc_)
+      MIOPEN_RETURN_IF_ERROR(miopenCreatePoolingDescriptor(&desc_));
+
+    int rank = gsl::narrow_cast<int>(kernel_shape.size());
+    InlinedVector<int> window(rank);
+    InlinedVector<int> padding(rank);
+    InlinedVector<int> stride(rank);
+    for (int i = 0; i < rank; i++) {
+      window[i] = gsl::narrow_cast<int>(kernel_shape[i]);
+    }
+    for (int i = 0; i < rank; i++) {
+      padding[i] = gsl::narrow_cast<int>(pads[i]);
+    }
+    for (int i = 0; i < rank; i++) {
+      stride[i] = gsl::narrow_cast<int>(strides[i]);
+    }
+    MIOPEN_RETURN_IF_ERROR(SetPoolingNdDescriptorHelper(
+        desc_,
+        mode,
+        MIOPEN_PROPAGATE_NAN,
+        rank,
+        window.data(),
+        padding.data(),
+        stride.data()));
+
+    return Status::OK();
+  }
+
+  operator miopenPoolingDescriptor_t() const { return desc_; }
+
+ private:
+  miopenPoolingDescriptor_t desc_;
+};
+
+template <typename T, typename PoolType>
+Status Pool<T, PoolType>::ComputeInternal(OpKernelContext* context) const {
+  typedef typename ToHipType<T>::MappedType HipT;
+  const Tensor* X = context->Input<Tensor>(0);
+  const TensorShape& x_shape = X->Shape();
+  const auto x_dims = x_shape.GetDims();
+
+  if (x_shape.NumDimensions() < 3) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Input dimension cannot be less than 3.");
+  }
+
+  auto kernel_shape = pool_attrs_.kernel_shape;
+  auto pads = pool_attrs_.pads;
+  auto strides = pool_attrs_.strides;
+
+  if (pool_attrs_.global_pooling) {
+    ORT_ENFORCE("unreachable");
```

Comment:
ORT_ENFORCE need a condition. may be ORT_ENFORCE(!pool_attrs_.global_pooling, '....')?